docs/source/hooks.rst

```diff
@@ -0,0 +1,173 @@
+.. _hooks:
+
+=====
+Hooks
+=====
+
+.. _hooks_introduction:
+
+Introduction
+============
+
+One of the goals of the **signac** framework is to make it easy to track the provenance of research data and to ensure its reproducibility.
+Hooks make it possible to track state changes to each job in a **signac** project as the :ref:`FlowProject<flow-project>` operates on it.
+
+A hook is a function that is called at a specific point relative to the execution of a **signac-flow** :ref:`operation<operations>`.
+A hook is triggered when an operation starts, exits, succeeds, or raises an exception.
+
+A basic use case is to log the success/failure of an operation by creating a hook that sets a job document value ``job.doc.operation_success`` to ``True`` or ``False``.
+Additionally, a user may record the `git commit ID <https://git-scm.com/book/en/v2/Git-Basics-Viewing-the-Commit-History>`_ upon the start of an operation, allowing them to track which version of code ran the operation.
+
+.. _hook_triggers:
+
+Triggers
+=============
+
+The following triggers are provided:
+
+1. :py:meth:`~flow.FlowProject.operation_hooks.on_start` will execute when the operation begins execution.
+2. :py:meth:`~flow.FlowProject.operation_hooks.on_exit` will execute when the operation exits, with or without error.
+3. :py:meth:`~flow.FlowProject.operation_hooks.on_success` will execute when the operation exits without error.
+4. :py:meth:`~flow.FlowProject.operation_hooks.on_exception` will execute when the operation exits with error.
+
+Hooks can be installed at the :ref:`operation level <operation-hooks>` or at the :ref:`flow-project level<project-level-hooks>`.
+Project-level hooks are called for every operation in the flow project.
+
+The hooks created with triggers :py:meth:`~flow.FlowProject.operation_hooks.on_start`, :py:meth:`~flow.FlowProject.operation_hooks.on_exit`,  and :py:meth:`~flow.FlowProject.operation_hooks.on_success` require two arguments: the operation name and the :py:class:`signac.contrib.job.Job` object. Hooks created to trigger :py:meth:`~flow.FlowProject.operation_hooks.on_exception` require three arguments: the operation name, the output error, and the job object.
```

Comment:
```suggestion
The hooks created with triggers :py:meth:`~flow.FlowProject.operation_hooks.on_start`, :py:meth:`~flow.FlowProject.operation_hooks.on_exit`, and :py:meth:`~flow.FlowProject.operation_hooks.on_success` require two arguments: the operation name and the :py:class:`signac.contrib.job.Job` object.
Hooks created to trigger :py:meth:`~flow.FlowProject.operation_hooks.on_exception` require three arguments: the operation name, the output error, and the job object.
```

Reply:
How are aggregates handled?

Reply:
See comment here where I got aggregation to work. https://github.com/glotzerlab/signac-docs/pull/154#discussion_r825338794

Not sure how to test it or describe it fully.